plugins/utils_operators/agol_operator.py

```diff
@@ -0,0 +1,92 @@
+#!python3
+# agol_operator.py - logic for all AGOL airflow operators
+
+import logging
+import os
+import json
+import hashlib
+from datetime import datetime
+from pathlib import Path
+
+import requests
+from airflow.models.baseoperator import BaseOperator
+from utils_operators.file_operator import DownloadFileOperator
+from airflow.utils.decorators import apply_defaults
+from utils import agol_utils
+
+
+class AGOLDownloadFileOperator(BaseOperator):
+    """
+    Downloads file from AGOL URL and saves to provided directory using provided filename.
+    This will always overwrite an existing file, if it's there.
+    This is because we are not interested in versioning this data - only pulling the latest from AGOL.
+
+    Returns a dictionary containing:
+        - path: path to saved file
+        - last_modified: timestamp file was last_modified (from the request)
+        - checksum: using md5 algorithm
+    """
+
+    @apply_defaults
+    def __init__(
+        self,
+        file_url: str,
+        dir: str,        
+        filename: str,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.file_url = file_url
```

Comment:
Could we update the logic here, and in the utils, so the `file_url` does not include the querystring parameters? It's a lot cleaner when building the DAG, and they are always the same. For example, the operator would only take:

`https://services.arcgis.com/S9th0jAJ7bqgIRjw/arcgis/rest/services/Reported_Crimes_ASR_RC_TBL_001/FeatureServer/0`

It would then build both calls by adding querystring parameters:

1. Fields:
```
{
        "where": "1=1",
        "outFields": "*",
        "outSR": 4326,
        "f": "json",
        "resultRecordCount": 1
    }
```

2. Features:
```
query_string_params = {
        "where": "1=1",
        "outFields": "*",
        "outSR": 4326,
        "f": "geojson",
        "resultType": "standard",
        "resultOffset": offset,
    }
```

All the logic is here (and it's better than the utils, more modular and with better coverage): https://github.com/open-data-toronto/snippets/blob/main/agol_endpoint_to_ckan.ipynb

Those functions could go into this operator. Once it's working I don't think we'd need the utils (we'd only use them for this operator).

So it would clean things up a lot.

Reply:
Hey @careduz I'm hesitant to do this. 

Between the URL im currently using and the neighborhoods URL, there are differences in the query parameters. I worry there could be more differences, and it would be hard to maintain an intelligent list of hardcoded query parameters in our utils or operators.

Since we dont control AGOL, I'd argue that we want to allow the developer using the operator to put in any URL they want and control the parameters themselves.

Let me know what you think here

Reply:
You know what - I'm going to put it in for now. I realize dont know enough about AGOL to make this claim, so pretend I said nothing.

We can always address it later if it comes up, but for now I'm going to use your logic from the ipynb in the operator

Reply:
No man, this is a very good point - let's have a chat, list the different parameters you are seeing and dig to the bottom of this. As people smarter than I say: "Trust but verify"